src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala

```diff
@@ -241,68 +243,76 @@ abstract class BCodeHelpers extends BCodeIdiomatic {
 
   /*
    * must-single-thread
+   *
+   * TODO: make this next claim true, if possible
+   *   by generating valid main methods as static in module classes
+   *   not sure what the jvm allows here
+   * + "  You can still run the program by calling it as " + sym.javaSimpleName + " instead."
    */
   object isJavaEntryPoint {
 
     /*
      * must-single-thread
      */
-    def apply(sym: Symbol, csymCompUnit: CompilationUnit): Boolean = {
-      def fail(msg: String, pos: Position = sym.pos) = {
-        reporter.warning(sym.pos,
-          sym.name +
-          s" has a main method with parameter type Array[String], but ${sym.fullName('.')} will not be a runnable program.\n  Reason: $msg"
-          // TODO: make this next claim true, if possible
-          //   by generating valid main methods as static in module classes
-          //   not sure what the jvm allows here
-          // + "  You can still run the program by calling it as " + sym.javaSimpleName + " instead."
-        )
-        false
-      }
-      def failNoForwarder(msg: String) = {
-        fail(s"$msg, which means no static forwarder can be generated.\n")
-      }
-      val possibles = if (sym.hasModuleFlag) (sym.tpe nonPrivateMember nme.main).alternatives else Nil
-      val hasApproximate = possibles exists { m =>
-        m.info match {
-          case MethodType(p :: Nil, _) => p.tpe.typeSymbol == definitions.ArrayClass
-          case _                       => false
-        }
-      }
-      // At this point it's a module with a main-looking method, so either succeed or warn that it isn't.
-      hasApproximate && {
-        // Before erasure so we can identify generic mains.
-        enteringErasure {
-          val companion     = sym.linkedClassOfClass
-
-          if (definitions.hasJavaMainMethod(companion))
-            failNoForwarder("companion contains its own main method")
+    def apply(sym: Symbol, csymCompUnit: CompilationUnit, mainClass: Option[String]): Boolean = sym.hasModuleFlag && {
+      val warn = mainClass.fold(true)(_ == sym.fullNameString)
+      def warnBadMain(msg: String, pos: Position): Unit = if (warn) reporter.warning(pos,
+        s"""|not a valid main method for ${sym.fullName('.')},
+            |  because $msg.
+            |  To define an entry point, please define the main method as:
+            |    def main(args: Array[String]): Unit
+            |""".stripMargin
+      )
+      def warnNoForwarder(msg: String, hasExact: Boolean, mainly: Type) = if (warn) reporter.warning(sym.pos,
+        s"""|${sym.name.decoded} has a ${if (hasExact) "valid " else ""}main method${if (mainly != NoType) " "+mainly else ""},
+            |  but ${sym.fullName('.')} will not have an entry point on the JVM.
+            |  Reason: $msg, which means no static forwarder can be generated.
```

Comment:
Open question: Do we want to bother people with static forwarders? Do people know what static forwarders are? Should they? Is 

> main1.scala:5: warning: Foo has a valid main method (args: Array[String])Unit,
  but foo1.Foo will not have an entry point on the JVM.
  Reason: companion is a trait, which means no static forwarder can be generated.

more useful to a user than 

> main1.scala:5: warning: Foo has a valid main method (args: Array[String])Unit,
  but foo1.Foo will not have an entry point on the JVM.
  Reason: companion is a trait

Reply:
There's the old cartoon where the dog owner is chatting with his dog, and what the dogs hears is, "Blah blah blah, Fido, blah blah blah."

I don't think anyone will say, You had me until static forwarder. Instead, they will read, `companion is a trait, blah blah blah`. Or maybe the don't know what a companion is, and they will understand "companion is a trait" to mean, "if I am alone, then I must be a bad person".

Reply:
"static forwarder" is google-able and therefore it's helpful, I think. it's also ignorable.

Reply:
I was hoping Seth would say, Oh I love The Far Side.

Reply:
blah blah blah blah blah opportunity to bikeshed error message wording blah blah blah blah

Reply:
It's even funnier and more relevant than I remembered:

https://www.flickr.com/photos/sluggerotoole/153603564/

The dog is getting a warning! Also, her name is Ginger.

It's also funny because "companion dog".